src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.cpp

```diff
@@ -90,52 +134,324 @@ namespace
             const ShadingPoint&     shading_point,
             ShadingComponents&      radiance) override      // output radiance, in W.sr^-1.m^-2
         {
-            PathVisitor light_path_visitor;
-            VolumeVisitor volume_visitor;
+            BDPTVertex camera_vertices[9];
+            BDPTVertex light_vertices[9];
 
-            PathTracer<PathVisitor, VolumeVisitor, true> light_path_tracer(     // true = adjoint
-                light_path_visitor,
-                volume_visitor,
-                ~0,
-                1,
-                ~0,
-                ~0,
-                ~0,
-                ~0,
-                shading_context.get_max_iterations());
+            size_t num_light_vertices = trace_light(sampling_context, shading_context, light_vertices);
+            size_t num_camera_vertices = trace_camera(sampling_context, shading_context, shading_point, camera_vertices);
+
+            for (size_t s = 0; s < num_light_vertices + 1; s++)
+                for (size_t t = 2; t < num_camera_vertices + 2; t++)
+                    if (s + t <= num_max_vertices)
+                        connect(shading_context, shading_point, light_vertices, camera_vertices, s, t, radiance);
+
+            shading_context.get_arena().clear();
```

Comment:
`shading_context.get_arena()` returns the per-pixel-sample arena. I don't think it should be cleared anywhere but in `GenericSampleRenderer::render_sample()`.

In fact, even the clearing of that arena in `pathtracer.h` looks suspicious to me (it was introduced by myself but I think we need to revisit this decision. `PathTracer` might want to provide a separate per-path-vertex arena instead).

Reply:
If I remember correctly, the code was there since the beginning of my development but I'll remove according to your request.